issue 4552 - Backup Redesign phase 3b - use dbimpl in replicatin plugin

progier389 opened the pull request:
Change description:
This replace Berkeley database dependencies in replication plugin
dbimpl API (as defined in https://directory.fedoraproject.org/docs/389ds/design/backend-redesign-phase3.html )
 is now used.
 
 Code impacted  is:
- the changelog and changelog cache (to use dbimp) 
- dbimpl and bdb (to add a few missing callbacks (txn handling) )
- dblayer (needed to move dblayer_strerror in dbimp to avoid getting in trouble with backend include in replication plugin ) 

relates: 
   #4469 - Backend redesign phase 3a - (dbimpl API implementation + bdb dependencies removal in back-ldbm)
   #4552 -  Backend redesign phase 3b - bdb dependencies removal in replication plugins 

Reviewed by: @Firstyear  @mreynolds389 

Platforms tested: F32

progier389 committed:
issue 4552 - Backup Redesign phase 3b - use dbimpl in replicatin plugin

Merge of a fix in cl5_clcache.c (changelog cache restarts from begining if large update)
Rebase with master

progier389 commented on ldap/servers/slapd/back-ldbm/dblayer.c:
```diff
@@ -1212,34 +1212,6 @@ dblayer_get_instance_data_dir(backend *be)
-        default:
-            return "Unexpected error code.";
-    }
-}
```
Note: dblayer_strerror will be rewritten in phase 3c (to build a map include file from dbimpl.h with mkDBErrStrs.py)

progier389 committed:
Issue 4469 - Backend redesing phase 3a - implement dbimpl API and use it in back-ldbm - fix test_maxbersize_repl pytest failure

mreynolds389 commented on ldap/servers/plugins/replication/cl5_clcache.c:
```diff
@@ -247,7 +239,14 @@ clcache_get_buffer(CLC_Buffer **buf, DB *db, ReplicaId consumer_rid, const RUV *
+                /* buf_busy_list is now set, and we can get the backend. So lets initialize the dbimpl buffers */
+                
+                 dblayer_bulk_set_buffer(be, &(*buf)->buf_bulk, &(*buf)->buf_bulkdata,
+                         WORK_CLC_BUFFER_PAGE_SIZE, DBI_VF_BULK_RECORD);
```
Indentation :-)

mreynolds389 commented on ldap/servers/plugins/replication/cl5_clcache.c:
```diff
@@ -376,16 +369,16 @@ clcache_load_buffer(CLC_Buffer *buf, CSN **anchorCSN, int *continue_on_miss, cha
                           (char *)buf->buf_key.data,
-                          flag == DB_NEXT ? "DB_NEXT" : "DB_SET");
-            rc = clcache_load_buffer_bulk(buf, DB_SET_RANGE);
+                          dblayer_op2str(dbop));
```
indentation 2

progier389 replied:
I do not see any indentation problem here. Can you precise  ?

mreynolds389 replied:
dblayer_op2str() is very far to the right.  It should be aligned with the "rc =" :-)

progier389 replied:
I think that alignment is correct:
dblayer_op2str is not the start of an instruction but the last parameter of slapi_log_err call
    It is aligned with the other parameters (as flag == DB_NEXT ?  ... was)

mreynolds389 replied:
Oh, you're right.  My bad...

mreynolds389 commented on ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c:
```diff
@@ -6401,7 +6401,10 @@ int bdb_public_bulk_nextrecord(dbi_bulk_t *bulkdata, dbi_val_t *key, dbi_val_t *
+    if (retdata == NULL || bulkdata->be == NULL) {
+        return DBI_RC_NOTFOUND;
+    }
+   return DBI_RC_SUCCESS;
```
indentation 3

mreynolds389 commented on ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c:
```diff
@@ -6666,6 +6670,42 @@ bdb_public_set_dup_cmp_fn(struct attrinfo *a, dbi_dup_cmp_t idx)
+{
+   return TXN_ABORT(txn);
+}
+
```
indentation for the above three functions...

mreynolds389 commented on ldap/servers/slapd/back-ldbm/dbimpl.c:
```diff
@@ -293,6 +298,33 @@ int dblayer_set_dup_cmp_fn(Slapi_Backend *be, struct attrinfo *a, dbi_dup_cmp_t
+            return "Database operation error: Transient error. transaction should be retried.";
+        case DBI_RC_OTHER:
+           return "Database operation error: Unhandled code. See details in previous error messages.";
+        default:
```
indentation around these returns

progier389 committed:
issue 4552 - Backup Redesign phase 3b - use dbimpl in replicatin plugin - fix indent issue

progier389 committed:
issue 4552 - Backup Redesign phase 3b - use dbimpl in replicatin plugin - fix merge issue

manual Merge of fix about changelog cache iteration restarting from beginning in case of large update + automatic rebase to master

Firstyear commented on ldap/servers/plugins/replication/cl5_api.c:
```diff
@@ -2365,23 +2352,23 @@ _cl5PurgeGetNextEntry(CL5Entry *entry, void *iterator, DBT *key)
         return rc;
     }
-    slapi_ch_free(&(data.data));
+    dblayer_value_free(it->it_cldb->be, &data);
```
Not a complaint, just a question about why the value free needs the be pointer now? Just want to know more about how this works :)

progier389 replied:
data is a dbi_val_t (i.e a struct) and as usual for struct, you need to refer it through its address ...
 FYI: I do not simply free data.data because data.flags determine whether the buffer need to be free or not.

Firstyear replied:
Right I see. So in this case &data is informing if the dblayer_value_free should free the be pointer?

progier389 replied:
If by "the pointer" you means data,data then yes 

int dblayer_value_free(Slapi_Backend *be __attribute__((unused)), dbi_val_t *data)
{   
    if (data != NULL && !(data->flags & DBI_VF_PROTECTED)) {
        slapi_ch_free(&data->data);
        data->size = data->ulen = 0;
    }
    return DBI_RC_SUCCESS;
}

mreynolds389 commented on ldap/servers/slapd/back-ldbm/dbimpl.c:
```diff
@@ -302,3 +334,65 @@ int dblayer_cursor_get_count(dbi_cursor_t *cursor, dbi_recno_t *count)
+                                     * data greater or equal than specified data
+                                     * then get the record.
+                                     */
+        "DBI_OP_MOVE_TO_RECNO",       /* move cursor to specified record number */
```
The comment indentation above it off.  Sorry :)

progier389 committed:
Issue 4552 - Backend redesign phase 3b - fix indent issue + random crash and memory leak in tombstone handling
